lib/ansible/executor/process/worker.py

```diff
@@ -35,18 +36,219 @@
+                    play_context.post_validate(templar=templar)
+
+                    # now that the play context is finalized, if the remote_addr is not set
+                    # default to using the host's address field as the remote address
+                    if not play_context.remote_addr:
+                        play_context.remote_addr = host.address
+
+                    # We also add "magic" variables back into the variables dict to make sure
+                    # a certain subset of variables exist.
+                    play_context.update_vars(task_vars)
+                except AnsibleError as e:
+                    context_validation_error = e
+
+                # FIXME remove code from TaskExecutor._execute() that is being moved into skip_worker
+                #       but keep/move it into TaskExecutor._run_loop() to account for loops
+                skip_worker(host, task, templar, task_vars, play_context)
+
+                if context_validation_error is not None:
+                    raise context_validation_error  # pylint: disable=raising-bad-type
+            else:
+                # FIXME move TaskExecutor._get_loop_items here? need to pass item to TE if non-empty
+                # FIXME if the cond is trivial and independent on loop vars, can we short-circuit?
+                # FIXME process loop here pre-fork if the task is include?
+                #       1. process when per item
+                #       2. process loop vars per item
+                #       3. create a result per item
+                #       4. send result per time
+                #       5. send ALL results
+                pass
+        except WorkerShortCircuit as short_circuit:
+            display.debug("skipping creating a worker for %s/%s" % (host, task))
+            display.debug("sending task result for task %s" % task._uuid)
+            self._tqm.send_callback('v2_runner_on_start', host, task)
+            self._final_q.send_task_result(
+                host.name,
+                task._uuid,
+                short_circuit.task_result,
+                task_fields=task.dump_attrs(),
+            )
+            display.debug("done sending task result for task %s" % task._uuid)
+        except AnsibleError as e:
+            # This is basically a copy of how TaskExecutor#run handles this.
+            # We could very well throw an AnsibleError in skip_worker() logic.
+            # We want to handle that so that things like ignore_errors=true
+            # still work.
+            res = {
+                'failed': True,
+                'msg': wrap_var(to_text(e, nonstring='simplerepr')),
+                '_ansible_no_log': play_context.no_log,
+            }
+            self._final_q.send_task_result(
+                host.name,
+                task._uuid,
+                res,
+                task_fields=task.dump_attrs())
+        # TODO: TaskExecutor#run handles plain Exception too, should we do that?
```

Comment:
Yeah we need to handle all exceptions that would be otherwise handled in `TaskExecutor` (`_execute()`, `run()`) and `Worker.run()`.